### PR TITLE
Fixes #12038 - adds pre-upgrade script for 2.4 to 3.0

### PR DESCRIPTION
**Pull request by jlsherrill**
this allows users to see what content hosts will be deleted
prior to running the upgrade

### CODE REVIEW

**Comment by jlsherrill** [comment]
note tests will not pass

**Comment by jlsherrill** [comment]
ALSO note this is for 2.4

**Comment by ehelms** [comment]
Will there also be one for 2.3?

File: lib/katello/tasks/preupgrade_content_host_check.rake [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+    def unregister_system(system, cause)
+      @systems_to_unregister[system] ||= []
+      @systems_to_unregister[system] << cause
+    end
+
+    def human_output(systems)
+      puts "\n\nSummary:\n"
+      puts "Content hosts to be preserved: #{::Katello::System.count - systems.count}"
+      puts "Content hosts to be delete: #{systems.count}"
```
Comment (waldenraines) [inline lib/katello/tasks/preupgrade_content_host_check.rake]: Minor typo: "deleted"

File: lib/katello/tasks/preupgrade_content_host_check.rake [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+      hostname = system.facts['network.hostname']
+
+      if hostname.nil?
+        unregister_system(system, "Missing hostname information.")
+        break
+      end
+
+      hosts = ::Host.where(:name => hostname)
+      if hosts.any? && hosts.where(:organization_id => system.environment.organization.id).empty? # host is not in the correct org
+        unregister_system(system, "Organization mis-match between Host (#{hosts[0].org.name}) and Content Host #{system.environment.organization.name}.")
```
Comment (waldenraines) [inline lib/katello/tasks/preupgrade_content_host_check.rake]: I think "mismatch" is one word.

**Comment by waldenraines** [comment]
LGTM and worked for everything I tested but since I wrote the original code that much of this is based on I will let someone else review/ACK this.

**Comment by jlsherrill** [comment]
@ehelms will open a 2.3 PR once this is merged

**Comment by jlsherrill** [comment]
@waldenraines updated

File: lib/katello/tasks/preupgrade_content_host_check.rake [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+    def csv_output(systems)
+      header = 'uuid, name, hostname, last_checkin, causes for deletion'
+      ouptut_lines = systems.map do |system, causes|
+        "#{system.uuid},#{system.name},#{system.hostname},#{system.lastCheckin},#{causes.join("  ")}"
+      end
+      "#{header}\n#{ouptut_lines.join("\n")}"
+    end
+
+    def write_csv(systems)
```
Comment (ehelms) [inline lib/katello/tasks/preupgrade_content_host_check.rake]: Why CSV?

Comment (jlsherrill) [inline lib/katello/tasks/preupgrade_content_host_check.rake]: wanted something that admins could review easily, other thoughts?

Comment (ehelms) [inline lib/katello/tasks/preupgrade_content_host_check.rake]: For importing into a spreadsheet? I guess that makes sense, if it were me I'd want JSON for that, but I suppose CSV can be parsed.

Comment (jlsherrill) [inline lib/katello/tasks/preupgrade_content_host_check.rake]: Right, if it were me i'd want json too, but I'm a developer and I think that sys admins typically perfer CSV

Comment (stbenjam) [inline lib/katello/tasks/preupgrade_content_host_check.rake]: +1

Could do YAML too, which is more human readable. But if there are a ton of systems, maybe CSV is easier for a sysadmin who wants to pull into LibreOffice and sort/search/etc on it.

**Comment by stbenjam** [comment]
@jlsherrill Can you touch a file somewhere to let us know this script was run? And bail out in the 3.0 rpm spec %pre if it doesn't exist - perhaps telling the user to run it?

We need something like that to address http://projects.theforeman.org/issues/14306 / https://bugzilla.redhat.com/show_bug.cgi?id=1316309

**Comment by jlsherrill** [comment]
@stbenjam added

**Comment by ehelms** [comment]
ACK - @stbenjam is your concern addressed?

**Comment by stbenjam** [comment]
Oh yup, sorry I missed the update.  My concern is addressed, APJ from me

**Pull request merged**